tests/unit/s2n_client_psk_extension_test.c

```diff
@@ -730,50 +600,111 @@ int main(int argc, char **argv)
         }
 
-        /* Receive a list with a match using customer's callback function */
+        /* Customer selection logic: customer selects out of bounds index */
         {
+            struct s2n_config *config = s2n_config_new();
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_set_psk_selection_callback(config, s2n_test_select_psk_identity_callback));
+
             struct s2n_connection *conn;
             EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
-
-            EXPECT_SUCCESS(s2n_config_set_psk_selection_callback(conn, s2n_test_select_psk_identity_callback));
-            EXPECT_EQUAL(conn->config->psk_selection_cb, s2n_test_select_psk_identity_callback);
-
-            struct s2n_stuffer wire_identities_in = { 0 };
-            EXPECT_SUCCESS(s2n_stuffer_alloc(&wire_identities_in, sizeof(wire_identities)));
-            EXPECT_SUCCESS(s2n_stuffer_write_bytes(&wire_identities_in, wire_identities, sizeof(wire_identities)));
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
 
             struct s2n_psk *match_psk = NULL;
             EXPECT_OK(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &match_psk));
             EXPECT_OK(s2n_psk_init(match_psk, S2N_PSK_TYPE_EXTERNAL));
             EXPECT_SUCCESS(s2n_psk_set_identity(match_psk, test_bytes_data, sizeof(test_bytes_data)));
 
-            EXPECT_OK(s2n_client_psk_recv_identity_list(conn, &wire_identities_in));
-            EXPECT_EQUAL(conn->psk_params.chosen_psk_wire_index, wire_identities_match_index);
-            EXPECT_EQUAL(conn->psk_params.chosen_psk, match_psk);
+            struct s2n_stuffer wire_identities_in = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&wire_identities_in, 0));
+            EXPECT_OK(s2n_write_test_identity(&wire_identities_in, &match_psk->identity));
+
+            s2n_test_customer_wire_index_choice = 10;
+            EXPECT_ERROR(s2n_client_psk_recv_identity_list(conn, &wire_identities_in));
+            EXPECT_EQUAL(conn->psk_params.chosen_psk, NULL);
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_config_free(config));
+            EXPECT_SUCCESS(s2n_stuffer_free(&wire_identities_in));
+        }
+
+        /* Customer selection logic: customer selects index without a local match */
+        {
+            struct s2n_config *config = s2n_config_new();
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_set_psk_selection_callback(config, s2n_test_select_psk_identity_callback));
+
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+            struct s2n_psk *local_psk = NULL;
+            EXPECT_OK(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &local_psk));
+            EXPECT_OK(s2n_psk_init(local_psk, S2N_PSK_TYPE_EXTERNAL));
```

Comment:
Instead of using `s2n_array_pushback` and `s2n_psk_init` we could use the `s2n_external_psk_new();` and `s2n_connection_append_psk()` APIs now that it is available.

Reply:
It doesn't really matter in this case, but for most unit tests s2n_connection_append_psk isn't preferred because it copies rather than inserts, so you can't then compare the psk pointer-to-pointer. Like, I couldn't say chosen_psk == local_psk if checking for a match. You have to compare the identities.

For consistency with the unit tests where it does matter, I'd prefer to just use pushback.